### PR TITLE
Make Tabulated3D EOS PuPable

### LINKED ISSUE #5420: Tabulated EoS Interpolator Needs to be PuP'd
# Bug reports:

### Expected behavior:

Executables using tabulated EoS run on execution.

### Current behavior:

Code would crash at runtime, during initialization.

### Environment:

Add as an attachment `$SPECTRE_BUILD_DIR/BuildInfo.txt` or
add its contents here.

# Feature request:

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Desired feature:

- [ ] Detail 1
- [ ] Detail 2
- [ ] Detail 3

### Detailed discussion:

Thanks to @nilsvu, a temporary fix has been proposed in Tabulated3d.cpp/hpp at 
https://github.com/sxs-collaboration/spectre/commit/12a4f8a924fa51504b2d9b252445627f4af63dd8#diff-ddaae7dca601b96e8a65295436a5f5919891980e9c7adf782c49851ce7124350L203

### PR DESCRIPTION
**Pull request by ermost**
Fix bug in Tabulated3D EOS pup function and add test.

### CODE REVIEW

**Comment by ermost** [comment]
Addresses #5420

File: src/PointwiseFunctions/Hydro/EquationsOfState/Tabulated3d.hpp [hunk 1]
```diff
@@ -121,6 +121,8 @@ class Tabulated3D : public EquationOfState<IsRelativistic, 3> {
                   std::vector<double> table_data, double energy_shift,
                   double enthalpy_minimum);
 
+  void initialize_interpolator();
```
Comment (nilsvu) [inline src/PointwiseFunctions/Hydro/EquationsOfState/Tabulated3d.hpp]: make this `private` please

File: tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Test_Tabulated3D.cpp [hunk 2]
```diff
@@ -239,4 +240,21 @@ SPECTRE_TEST_CASE("Unit.PointwiseFunctions.EquationsOfState.Tabulated3D",
...
+          {"Tabulated3D:\n"
+           "  TableFilename: " +
+           unit_test_src_path() +
+           "PointwiseFunctions/Hydro/EquationsOfState/dd2_unit_test.h5\n"
+           "  TableSubFilename: 'dd2'"}));
+  const EoS::Tabulated3D<true>& deserialized_eos =
+      dynamic_cast<const EoS::Tabulated3D<true>&>(*eos_pointer);
+  TestHelpers::EquationsOfState::test_get_clone(deserialized_eos);
+
+  CHECK(deserialized_eos == eos);
```
Comment (nilsvu) [inline tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Test_Tabulated3D.cpp:263]: I think this added test would have succeeded before. Maybe just `serialize_and_deserialize` the `eos` in line 144 before all the checks? Up to you, just make sure the deserialized EOS performs an interpolation.

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Test_Tabulated3D.cpp (hunk 2 above)
Comment (ermost) [reply tests/Unit/PointwiseFunctions/Hydro/EquationsOfState/Test_Tabulated3D.cpp:263]: I now repeat the same interpolation call as above.

**Review by nilsvu — DISMISSED** [review]

**Comment by nilsvu** [comment]
Clang-tidy has some complaints. I pushed a small fix to two of them. Can't do anything about the complaint in Boost.

**Review by nilsvu — DISMISSED** [review]

**Review by nilsvu — APPROVED** [review]

**Review by pajkosmi — APPROVED** [review]

**Pull request merged**